File: 3488-closest-equal-element-queries/3488-closest-equal-element-queries.py
class Solution:
    def solveQueries(self, nums: List[int], queries: List[int]) -> List[int]:
        him={}
        n=len(nums)
        for i in range(n):
            if nums[i] not in him:
                him[nums[i]]=[]
            him[nums[i]].append(i)
        ans =[]
        for i in range(len(queries)):
            ele=nums[queries[i]]
            l=him[ele]
            if len(l)==1:
                ans.append(-1)
                continue
            ind=self.find(l,queries[i])
            left= abs(queries[i]- l[((ind-1+len(l))%len(l))])
            right=abs(queries[i]- l[((ind+1)%len(l))])
            leftbacha=n-left
            rightbacha=n-right
            ans.append(min(left,right,leftbacha,rightbacha))
        return ans
    def find(self,a,t):
        s=0
        r=len(a)-1
        while(s<=r):
            mid=s+(r-s)//2
            if a[mid]==t:
                return mid
            elif (t>=a[mid]):
                s=mid+1
            else:
                r=mid-1
        return -1


        

File: 3488-closest-equal-element-queries/3488-closest-equal-element-queries.java
class Solution {
    public List<Integer> solveQueries(int[] nums, int[] queries) {
    Map<Integer,List<Integer>> him=new HashMap<>();
    for(int i=0;i<nums.length;i++){
        if (!him.containsKey(nums[i])){
            him.put(nums[i],new ArrayList<>());
        }
        him.get(nums[i]).add(i);
    }
    List<Integer> ans=new ArrayList<>();
    for (int i=0;i<queries.length;i++){
        int ele=nums[queries[i]];
        List<Integer>all=him.get(ele);
        if (all.size()==1){
            ans.add(-1);
            continue;
        }
        int ind=find(all,queries[i]);
        // int ind=all.indexOf(queries[i]); giving time limit exceeded
        int left=all.get(((ind-1+all.size())%all.size()));
        int right=all.get(((ind+1)%all.size()));
        int diffleft=Math.abs(queries[i]-left);
        int bachaleft=Math.abs(nums.length-diffleft);
        int diffright=Math.abs(queries[i]-right);
        int bacharight=Math.abs(nums.length-diffright);
        int t=Math.min(diffleft,Math.min(bachaleft,Math.min(diffright,bacharight)));
            ans.add(t);
    }
    return ans;
  
    }
    int find(List<Integer> l,int target){
        int s=0;
        int r=l.size()-1;
        while(s<=r){
            int mid=s+(r-s)/2;
            if (l.get(mid).equals(target)){
                return mid;
            }else if (l.get(mid)>target){
                r=mid-1;
            }else{
                s=mid+1;
            }
        }
        return -1;
    }
}